modules/dnn/src/layers/gather_layer.cpp

```diff
@@ -115,7 +102,7 @@ class GatherLayerImpl CV_FINAL : public GatherLayer
             const size_t src_offset = i * outer_size;
             for (size_t j = 0 ; j < indices_total; ++j)
             {
-                const int index = idx[j];
+                const int index = normalize_axis(idx[j], axis_size);
```

Comment:
As any type is casted to int anyway, I recommend to do before forward_impl. So it let's avoid using a template method.

Reply:
As I understand, OpenCV supports only int32 Mat sizes, so there is no point to keep int64 index value.
ConvertTo creates a copy of Mat, so it uses some memory and CPU resources. Casting of a local variable doesn't uses additional memory and most likely doesn't affect performance.